File: 0002-add-two-numbers/0002-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverse(ListNode head) {
        ListNode c=head, p=null, n=null;
        while(c!=null)
        {
            n=c.next;
            c.next=p;
            p=c;
            c=n;
        }
        return p;
    }
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode a = l1;
        ListNode b = l2;
        
        ListNode c = new ListNode(-1);
        ListNode curr = c;
        while(a!=null || b!=null)
        {
            int f = 0;
            if(a!=null){
                f = a.val;
                a = a.next;
            }
            int s = 0;
            if(b!=null){
                s = b.val;
                b = b.next;
            }
            curr.next = new ListNode(f+s);
            curr = curr.next;
        }
        
        curr = c.next;
        int carry = 0;
        while(curr!=null)
        {
            int temp = (carry+curr.val)/10;
            curr.val = (carry+curr.val)%10;
            carry = temp;
            if(curr.next == null && carry!=0)
            {
                curr.next = new ListNode(carry);
                break;
            }
            curr = curr.next;
        }
        return c.next;
    }
}

File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0)
        curr = dummy
        carry = 0
        
        while l1 or l2 or carry:
            sum_value = carry
            
            if l1:
                sum_value += l1.val
                l1 = l1.next
                
            if l2:
                sum_value += l2.val
                l2 = l2.next
            
            carry = sum_value // 10
            curr.next = ListNode(sum_value % 10)
            curr = curr.next
        
        return dummy.next
